src/sentry/scripts/similarity/index.lua

```diff
@@ -278,507 +370,334 @@ local function scale_to_total(values)
     return result
 end
 
-local function collect_index_key_pairs(arguments, validator)
-    return build_variadic_argument_parser({
-        {"index", identity},
-        {"key", identity},
-    }, validator)(arguments)
+
+-- Signature Matching
+
+local function pack_frequency_coordinate(band, bucket)
+    return struct.pack('>B', band) .. bucket
 end
 
+local function unpack_frequency_coordinate(field)
+    local band, index = struct.unpack('>B', field)
+    return band, string.sub(field, index)
+end
 
--- Signature Matching
+local function get_bucket_membership_set(configuration, index, band, bucket)
+    return TimeSeriesSet:new(
+        configuration.interval,
+        configuration.retention,
+        configuration.timestamp,
+        function (i)
+            return string.format(
+                '%s:m:%s:',
+                get_key_prefix(configuration, index),
+                i
+            ) .. pack_frequency_coordinate(band, bucket)
+        end
+    )
+end
 
-local function parse_band(configuration, arguments, cursor)
-    local result = {}
+local function get_frequencies(configuration, index, item)
+    local frequencies = {}
+    for i = 1, configuration.bands do
+        frequencies[i] = {}
+    end
 
-    local count = tonumber(arguments[cursor])
-    cursor = cursor + 1
+    local key = get_frequency_key(configuration, index, item)
+    local response = redis.call('HGETALL', key)
+    for field, value in redis_hash_response_iterator(response) do
+        local band, bucket = unpack_frequency_coordinate(field)
+        frequencies[band][bucket] = tonumber(value)
+    end
+
+    return frequencies
+end
+
+local function set_frequencies(configuration, index, item, frequencies)
+    local key = get_frequency_key(configuration, index, item)
 
-    for i = 1, count do
-        result[arguments[cursor]] = tonumber(arguments[cursor + 1])
-        cursor = cursor + 2
+    for band, buckets in ipairs(frequencies) do
+        for bucket, count in pairs(buckets) do
+            local field = pack_frequency_coordinate(band, bucket)
+            redis.call('HINCRBY', key, field, count)
+        end
     end
 
-    return result, cursor
+    -- TODO: This is less accurate than it could be at this point -- it should
+    -- probably just emulate the TTL behavior of the application.
```

Comment:
This can result in data existing in the index longer than it is actually stored in the database.

I also wonder if it might make sense to not actually trim the candidate list during search to allow handling things like frequency evictions more gracefully.